### PR TITLE
Telemetry integration

### PR DESCRIPTION
**Pull request by RyanSiu1995**
Issue #260 
Here is the prototyping for the telemetry middleware.
Please kindly have a comment on the functionality and flexibility of it.
Thank you!

### CODE REVIEW

File: lib/tesla/middleware/telemetry.ex [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+      :telemetry_app.start(nil, nil)
+      :telemetry.attach(
+        "tesla-telemetry",
+        [:tesla, :telemetry, :traffic],
+        fn ([:tesla, :telemetry, :traffic], time, meta, _config) -> # Do sth end,
+        nil)
+    end
+  end
+
+  defmodule MyClient do
```
Comment (teamon) [inline lib/tesla/middleware/telemetry.ex]: I'd put this definition on top and then only this lines:

```elixir
:telemetry.attach("my-tesla-telemetry", [:tesla, :telemetry, :traffic], fn event, time, meta, config ->
  # do something with the event
end)
```

And reference [telemetry](https://github.com/beam-telemetry/telemetry) as the source of documentation for further usage

File: lib/tesla/middleware/telemetry.ex [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    plug Tesla.Middleware.Telemetry
+
+  end
+  ```
+  """
+
+  @doc false
+  def call(env, next, _opts) do
+    {time, res} = :timer.tc(Tesla, :run, [env, next])
+    :telemetry.execute([:tesla, :telemetry, :traffic], time, Enum.into([res], %{}))
```
Comment (teamon) [inline lib/tesla/middleware/telemetry.ex]: I'm not sure about the event name, maybe we should use `[:tesla, :request, :done]` or even `[:tesla, :request]`? The `:telemetry` part seems redundant.

File: mix.exs [hunk 3]
```diff
@@ -68,6 +68,7 @@ defmodule Tesla.Mixfile do
 
       # other
       {:fuse, "~> 2.4", optional: true},
+      {:telemetry, "~> 0.3.0", optional: true},
```
Comment (teamon) [inline mix.exs]: I'd use `~> 0.3` #yolo ;)

File: lib/tesla/middleware/telemetry.ex (hunk 2 above)
Comment (teamon) [inline lib/tesla/middleware/telemetry.ex]: The `Enum.into(...)` part will result in:

```ex
%{ok: %Tesla.Env{status: 200, ...}}
```

or 

```ex
%{error: :nxdomain}
```

I would leave the result as is, like `%{result: {:ok, env}}` or `%{result: {:error, ...}}` to make it consistent (and you can still use nice pattern matching on this structure)

Comment (ericmj) [reply lib/tesla/middleware/telemetry.ex]: 👍 for `[:tesla, :request]`. `telemetry`, `traffic`, and `done` seems redundant.

**Comment by teamon** [comment]
@RyanSiu1995  can you please rebase over new master as I've updated travis config to fix the build.

**Pull request merged**